silverback/subscriptions.py

```diff
@@ -0,0 +1,150 @@
+
+        return response
+
+    def _create_request(self, method: str, params: list) -> dict:
+        self._last_request += 1
+        return {
+            "jsonrpc": "2.0",
+            "id": self._last_request,
+            "method": method,
+            "params": params,
+        }
+
+    async def _get_response(self, request_id: int) -> dict:
+        if buffer := self._rpc_msg_buffer:
+            for idx, data in enumerate(buffer):
+                if data.get("id") == request_id:
+                    self._rpc_msg_buffer.pop(idx)
+                    return data
+
+        async with self._ws_lock:
+            tries = 0
+            while tries < self.rpc_response_timeout_count:
+                if self._rpc_msg_buffer and self._rpc_msg_buffer[-1].get("id") == request_id:
+                    return self._rpc_msg_buffer.pop()
+
+                # NOTE: Python <3.10 does not support `anext` function
+                await self.__anext__()  # Keep pulling until we get a response
+
+        raise RuntimeError("Timeout waiting for response.")
+
+    async def subscribe(self, type: SubscriptionType, **filter_params) -> str:
+        if not self.connection:
+            raise ValueError("Connection required.")
+
+        if type is SubscriptionType.BLOCKS and filter_params:
+            raise ValueError("blocks subscription doesn't accept filter params.")
+
+        request = self._create_request(
+            "eth_subscribe",
+            [type.value, filter_params] if type is SubscriptionType.EVENTS else [type.value],
+        )
+        await self.connection.send(json.dumps(request))
+        response = await self._get_response(request.get("id") or self._last_request)
+
+        sub_id = response.get("result")
+        if not sub_id:
+            # NOTE: Re-dumping message to avoid type-checking concerns.
+            raise ValueError(f"Missing subscription ID in response: {json.dumps(response)}.")
+
+        return sub_id
+
+    async def get_subscription_data(self, sub_id: str) -> AsyncGenerator[dict, None]:
+        while True:
+            if not (queue := self._subscriptions.get(sub_id)) or queue.empty():
+                async with self._ws_lock:
+                    # Keep pulling until a message comes to process
+                    # NOTE: Python <3.10 does not support `anext` function
+                    await self.__anext__()
+            else:
+                yield await queue.get()
+
+    async def unsubscribe(self, sub_id: str) -> bool:
+        if sub_id not in self._subscriptions:
+            raise ValueError(f"Unknown sub_id '{sub_id}'")
+
+        if not self.connection:
+            # Nothing to unsubscribe.
+            return True
+
+        request = self._create_request("eth_unsubscribe", [sub_id])
+        await self.connection.send(json.dumps(request))
+
+        response = await self._get_response(request.get("id") or self._last_request)
+        if success := response.get("result", False):
```

Comment:
The walrus operator is strictly `None` checking, so the `False` default here will not trigger the `del ` statement.

```python
>>> fal = False
>>> if x:= fal:
...     print("yes")
...
>>>

Is that what you intend here? To only do said deletion on `None` and not `False`? I just wanna make sure
```

Reply:
sorry i confused myself

Reply:
so `x` still gets set to `False` but the statement doesn't run, interesting

Reply:
Yeah seems like the assignment only happens if not None, but the condition is evaluated normally